feat(linter): implement no-useless-return rule

FMP1783 opened the pull request:
Implement `eslint/no-useless-return` rule.

Disallows redundant return statements that have no effect on function behavior.

## Test

- 27 valid / 25 invalid test cases
- `just ready` passes

Note: `globalReturn` cases excluded (unsupported by oxc)

Part of #479

FMP1783 committed:
feat(linter): implement no-useless-return rule

Implement ESLint no-useless-return rule that disallows redundant return statements.

A return statement with nothing after it is redundant and has no effect on runtime behavior.

Part of #479

camc314 commented on crates/oxc_linter/src/rules/eslint/no_useless_return.rs:
```diff
@@ -0,0 +1,604 @@
+                    {
+                        return false;
+                    }
+                }
```
it this always true? or does eslint just not report on this case.

For example:
```ts
function foo(arg0) {
    if (arg0) {
        return "i'm truthy"
    } else {
        return
    }
}

camc314 committed:
add some more test cases

camc314 committed:
use `Span::contains_inclusive`

camc314 committed:
simplify case searching with `Iter::position`

camc314 committed:
remove needless node lookup

camc314 committed:
remove unreachable code.
since we early return when we see a function body, reaching an arrow function expressino/function is impossible

camc314 committed:
update test case comment

camc314 committed:
refactor(linter/no-useless-return): simplify implementation with single-pass ancestor analysis

camc314 reviewed (approved):
Amazing work, thank you!

camc314 committed:
update lintgen

camc314 committed:
lint
